### PR TITLE
Create SwarmPad object and create a CLI tool to offer its functionality as CLI commands

### PR DESCRIPTION
**Pull request by KalanaDananjaya**
- Resolves https://github.com/SwarmForm/SwarmForm/issues/17

- SwarmPad class inherits from LaunchPad class

- It overrides from_dict, reset, _restart_ids functions of Launchpad 

- Offers a CLI tool for handling the SwarmPad. Following commands are available
```
sform init
sform add 
sform reset 
sform get_sf
sform cluster
```

### CODE REVIEW

File: swarmform/core/swarmpad.py [hunk 1]
```diff
@@ -3,11 +3,39 @@
 from fireworks import LaunchPad, Firework
 from fireworks.core.launchpad import LazyFirework
 from fireworks.fw_config import GRIDFS_FALLBACK_COLLECTION
+from fireworks.utilities.fw_serializers import serialize_fw
+
```
Comment (AyeshW) [inline swarmform/core/swarmpad.py]: Unused import

File: swarmform/core/swarmpad.py [hunk 2]
```diff
@@ -184,20 +212,20 @@ def get_sf_by_id(self, sf_id):
 						 links_dict['metadata'], links_dict['created_on'],
 						 links_dict['updated_on'], None, links_dict['sf_id'])
 
-	def get_sf_by_name(self, wf_name):
+	def get_sf_by_name(self, sf_name):
```
Comment (AyeshW) [inline swarmform/core/swarmpad.py:164]: Better if we can comment what happen if there are multiple SwarmFlows with the same name

File: swarmform/setup.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+if __name__ == "__main__":
+    setup(
+        name='SwarmForm',
+        version="0.1",
+        description='SwarmForm Workflow Management Software',
+        long_description=open(os.path.join(module_dir, 'README.md')).read(),
+        url='https://github.com/SwarmForm/SwarmForm',
+        author='Kalana Wijethunga, Randika Jayasekara, Ayesh Weerasinghe',
+        author_email='kalana.16@cse.mrt.ac.lk, rpjayaseka.16@cse.mrt.ac.lk, ayeshw.16@cse.mrt.ac.lk',
```
Comment (AyeshW) [inline swarmform/setup.py]: Invalid emails. Ha haa

File: swarmform/examples/cluster_examples/ex_cluster.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+	unclustered_sf = SwarmFlow.from_file(filename)
+
+	# store workflow
+	swarmpad.add_sf(unclustered_sf)
+
+	sf = swarmpad.get_sf_by_id(unclustered_sf.sf_id)
+
+	# Cluster the SwarmFlow
+	clustered_sf = cluster_sf(swarmpad, unclustered_sf.sf_id)
+
```
Comment (AyeshW) [inline swarmform/examples/cluster_examples/ex_cluster.py:21]: Better to add the clustered SwarmFlow to the DB

File: swarmform/scripts/sform_run.py [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+    sys.stderr.write("Interrupted by signal {:d}\n".format(signum))
+    sys.exit(1)
+
+
+def get_output_func(format):
+    if format == "json":
+        return lambda x: json.dumps(x, default=DATETIME_HANDLER, indent=4)
+    else:
+        return lambda x: yaml.safe_dump(recursive_dict(x, preserve_unicode=False),
+                                        default_flow_style=False)
```
Comment (AyeshW) [inline swarmform/scripts/sform_run.py]: This function can be directly imported from _lpad_run_

File: swarmform/scripts/sform_run.py [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+            return SwarmPad.from_file(args.launchpad_file)
+        else:
+            args.loglvl = 'CRITICAL' if args.silencer else args.loglvl
+            return SwarmPad(logdir=args.logdir, strm_lvl=args.loglvl)
+    except Exception:
+        traceback.print_exc()
+        err_message = 'SwarmForm was not able to connect to MongoDB. Is the server running? ' \
+                      'The database file specified was {}.'.format(args.launchpad_file)
+        if not args.launchpad_file:
+            err_message += ' Type "lpad init" if you would like to set up a file that specifies ' \
```
Comment (AyeshW) [inline swarmform/scripts/sform_run.py]: Shouldn't it be changed as _sform init_ ?

File: swarmform/scripts/sform_run.py [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+            from fireworks.utilities.dagflow import DAGFlow
+            DAGFlow.from_fireworks(fwf)
+        sp.add_sf(fwf)
+
+
+def get_sf(args):
+    if args.id:
+        get_sf_by_id(args)
+    if args.name:
+        get_sf_by_name(args)
```
Comment (AyeshW) [inline swarmform/scripts/sform_run.py]: Don't we need to return these values?

File: swarmform/scripts/sform_run.py [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+                isinstance(args.fw_id, six.string_types):
+            if "," in args.fw_id:
+                args.fw_id = [int(x) for x in args.fw_id.split(",")]
+            else:
+                args.fw_id = [int(args.fw_id)]
+
+        args.func(args)
+
+    signal.signal(signal.SIGINT, handle_interrupt)  # graceful exit on ^C
+
```
Comment (AyeshW) [inline swarmform/scripts/sform_run.py:188]: Need one more blank line

File: swarmform/scripts/sform_run.py [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+    parser.add_argument('-o', '--output', choices=["json", "yaml"],
+                        default='json', type=lambda s: s.lower(),
+                        help='Set output display format to either json or YAML. '
+                             'YAML is easier to read for long documents. JSON is the default.')
+
+    parser.add_argument('-l', '--launchpad_file', help='path to LaunchPad file containing '
+                                                       'central DB connection info',
+                        default=None)
+    parser.add_argument('-c', '--config_dir',
+                        help='path to a directory containing the LaunchPad file (used if -l unspecified)',
```
Comment (AyeshW) [inline swarmform/scripts/sform_run.py:125]: Can we change here _launchpad_ and _LaunchPad_ to _swarmpad_ or is there any problem with that?

**Review by AyeshW — CHANGES_REQUESTED** [review]
- It is better if you can add required functions MUST be overridden or implemented in swarmpad and swarm_run later as a _TODO_ .

- As many of the SwarmPad and swarm_run functions are completely overridden and implemented just because of the return type of corresponding functions of Fireworks library is Workflow or LaunchPad, we can modify our constructors to create SwarmFlow and SwarmPad from a given Workflow or LaunchPad.

File: swarmform/scripts/sform_run.py (hunk 7 above)
Comment (KalanaDananjaya) [reply swarmform/scripts/sform_run.py]: These are only getting printed.No need to written anything. I will send an update with printing functionality

File: swarmform/scripts/sform_run.py (hunk 9 above)
Comment (KalanaDananjaya) [reply swarmform/scripts/sform_run.py:125]: I tried changing that and faced many issues. Better keep it this way until we want to make this production ready

**Review by AyeshW — COMMENTED** [review]
sf_id is missed after executing the clustered SwarmFlow. Need to check this out

**Review by AyeshW — APPROVED** [review]

**Pull request merged**